### PR TITLE
Emit the Luis Trace Info from the v3 C# SDK

### PR DESCRIPTION
**Pull request by emadelwany**

### CODE REVIEW

File: CSharp/Library/Microsoft.Bot.Builder/Dialogs/LuisDialog.cs [hunk 1]
```diff
@@ -140,6 +143,10 @@ public LuisServiceResult(LuisResult result, IntentRecommendation intent, ILuisSe
     [Serializable]
     public class LuisDialog<TResult> : IDialog<TResult>
     {
+        public const string LuisTraceType = "https://www.luis.ai/schemas/trace";
+        public const string LuisTraceLabel = "Luis Trace";
+        public const string Obfuscated = "****";
```
Comment (tomlm) [inline CSharp/Library/Microsoft.Bot.Builder/Dialogs/LuisDialog.cs]: >Obfuscated [](start = 28, length = 10)

????

File: CSharp/Library/Microsoft.Bot.Builder/Dialogs/LuisDialog.cs [hunk 2]
```diff
@@ -294,11 +311,24 @@ protected virtual Task<string> GetLuisQueryTextAsync(IDialogContext context, IMe
...
+
+        private static async Task EmitTraceInfo(IBotContext context, LuisResult luisResult, ILuisOptions luisOptions, ILuisModel luisModel)
+        {
+            var luisTraceInfo = new LuisTraceInfo
+            {
+                LuisResult = luisResult,
+                LuisOptions = luisOptions,
+                LuisModel = luisModel
+            };
+            var activity = Activity.CreateTraceActivityReply(context.Activity as Activity, "LuisDialog", LuisTraceType, luisTraceInfo, LuisTraceLabel) as IMessageActivity;
```
Comment (tomlm) [inline CSharp/Library/Microsoft.Bot.Builder/Dialogs/LuisDialog.cs]: >IMessageActivity [](start = 154, length = 16)

wow, that stoopid :)

File: CSharp/Library/Microsoft.Bot.Builder/Dialogs/LuisTraceInfo.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        [JsonProperty("luisResult")]
+        public LuisResult LuisResult { set; get; }
+
+        [JsonProperty("luisOptions")]
+        public ILuisOptions LuisOptions { get; set; }
+
+        [JsonProperty("luisModel")]
+        public ILuisModel LuisModel { get; set; }
+    }
```
Comment (tomlm) [inline CSharp/Library/Microsoft.Bot.Builder/Dialogs/LuisTraceInfo.cs]: comments?

Comment (emadelwany) [reply CSharp/Library/Microsoft.Bot.Builder/Dialogs/LuisTraceInfo.cs]: Added

**Comment by tomlm** [comment]
>             {

You need to add filter here too.

---
Refers to: CSharp/Library/Microsoft.Bot.Connector.Shared/ConnectorAPI/ConversationsExtensions.cs:227 in 480d918. [](commit_id = 480d91881118725edc210e97d10a7015f93b8851, deletion_comment = False)

File: CSharp/Library/Microsoft.Bot.Builder/Dialogs/LuisDialog.cs (hunk 2 above)
Comment (emadelwany) [reply CSharp/Library/Microsoft.Bot.Builder/Dialogs/LuisDialog.cs]: Agreed, v3 uses IMessageActivities every where and not IActivity, and is very "reply" focused as I told you earlier today :( Can you think of a better way of doing this?

File: CSharp/Library/Microsoft.Bot.Builder/Dialogs/LuisDialog.cs (hunk 1 above)
Comment (emadelwany) [reply CSharp/Library/Microsoft.Bot.Builder/Dialogs/LuisDialog.cs]: Added it but forgot to use it, fixed, thanks!

**Pull request merged**